src/java/za/co/mip/ablduck/ABLDuck.java

```diff
@@ -0,0 +1,463 @@
+                            Paths.get(this.destDir.getAbsolutePath(), file));
+                    replaceTemplateTags("{version}", Version.getPCTVersion(),
+                            Paths.get(this.destDir.getAbsolutePath(), file));
+                    replaceTemplateTags("{date}", formatter.format(new Date()),
+                            Paths.get(this.destDir.getAbsolutePath(), file));
+                }
+            } catch (IOException ex) {
+                throw new BuildException(ex);
+            }
+        }
+
+        Gson gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation().create();
+        DataJSObject dataJSObject = new DataJSObject();
+        HTMLGenerator html = new HTMLGenerator();
+        IPropath pp = new Propath(
+                new org.eclipse.core.runtime.Path(getProject().getBaseDir().getAbsolutePath()),
+                propath.list());
+        IASTContext astContext = new PropathASTContext(pp);
+        IProgressMonitor monitor = new NullProgressMonitor();
+        IASTManager astMgr = ASTManager.getASTManager();
+
+        log("Generating ABLDuck documentation (11.5+ method)", Project.MSG_INFO);
+
+        for (FileSet fs : filesets) {
+            // And get files from fileset
+            String[] dsfiles = fs.getDirectoryScanner(this.getProject()).getIncludedFiles();
+
+            for (int i = 0; i < dsfiles.length; i++) {
+                File file = new File(fs.getDir(this.getProject()), dsfiles[i]);
+                log("Generating AST for " + file.getAbsolutePath(), Project.MSG_VERBOSE);
+
+                int extPos = file.getName().lastIndexOf('.');
+                String ext = file.getName().substring(extPos);
+                boolean isClass = ".cls".equalsIgnoreCase(ext);
+
+                ICompilationUnit root = astMgr.createAST(file, astContext, monitor,
+                        IASTManager.EXPAND_ON, IASTManager.DLEVEL_FULL);
+                if (isClass) {
+                    ABLDuckClassVisitor visitor = new ABLDuckClassVisitor(pp, this);
+                    log("Executing AST ClassVisitor " + file.getAbsolutePath(),
+                            Project.MSG_VERBOSE);
+                    root.accept(visitor);
+
+                    try {
+                        SourceJSObject jsObject = visitor.getJSObject();
+                        jsObjects.put(jsObject.name, jsObject);
+                    } catch (IOException ex) {
+                        throw new BuildException(ex);
+                    }
+                }
+            }
+        }
+
+        // Determine class hierarchy, subclasses and search objects
+        for (Map.Entry<String, SourceJSObject> j : jsObjects.entrySet()) {
+            SourceJSObject js = j.getValue();
+
+            // Class tree
+            ClassDataObject cls = new ClassDataObject();
+            cls.name = js.name;
+            cls.ext = js.ext;
+            cls.icon = "icon-" + js.classIcon;

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "icon-" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1192)